src/audio/OpenALAudioEngine.cpp

```diff
@@ -0,0 +1,488 @@
+#ifdef RE_USE_SOUND
+
+#include "OpenALAudioEngine.h"
+
+#include <algorithm>
+#include <array>
+#include <atomic>
+#include <mutex>
+#include <thread>
+
+#include <AL/al.h>
+#include <AL/alc.h>
+
+using namespace Audio;
+
+static std::vector<std::string> enumeratedDevices;
+
+// The mutex is there so that static members
+// can be accessed in a thread-safe manner
+static std::mutex mtx_enumeratedDevices;
+
+constexpr std::size_t OPENAL_SOURCES_POOL = 256;
+
+std::string getErrorString(ALenum error)
+{
+    switch (error)
+    {
+        case AL_NO_ERROR:
+            return "No error";
+        case AL_INVALID_NAME:
+            return "An invalid id was passed to an OpenAL function";
+        case AL_INVALID_ENUM:
+            return "An invalid enum value was passed to an OpenAL function";
+        case AL_INVALID_VALUE:
+            return "An invalid value was passed to an OpenAL function";
+        case AL_INVALID_OPERATION:
+            return "The requested OpenAL operation is not valid";
+        case AL_OUT_OF_MEMORY:
+            return "The requested operation resulted in OpenAL running out of memory";
+        default:
+            return "Unknown error";
+    }
+}
+
+OpenALException::OpenALException(ALenum error)
+    : std::runtime_error(getErrorString(error))
+    , m_err(error)
+{
+}
+
+const std::vector<std::string>& OpenALAudioEngine::devices()
+{
+    std::lock_guard<std::mutex> lock(mtx_enumeratedDevices);
+
+    if (enumeratedDevices.empty())
+    {
+        if (alcIsExtensionPresent(nullptr, "ALC_ENUMERATION_EXT"))
+        {
+            size_t len = 0;
+            const ALCchar* devices = alcGetString(nullptr, ALC_DEVICE_SPECIFIER);
+            const ALCchar *device = devices, *next = devices + 1;
+
+            while (device && *device != '\0' && next && *next != '\0')
+            {
+                enumeratedDevices.push_back(device);
+
+                len = strlen(device);
+                device += (len + 1);
+                next += (len + 2);
+            }
+        }
+        else
+        {
+            enumeratedDevices.push_back("");
+        }
+    }
+
+    return enumeratedDevices;
+}
+
+OpenALAudioEngine::OpenALAudioEngine(const std::string& deviceName)
+{
+    m_device = alcOpenDevice(deviceName.empty() ? nullptr : deviceName.c_str());
+    if (!m_device) throw OpenALException(alGetError());
+
+    m_context = alcCreateContext(m_device, nullptr);
+    if (!m_context) throw OpenALException(alGetError());
+
+    captureContext();
+
+    m_freeSources.resize(OPENAL_SOURCES_POOL);
+    alGenSources(OPENAL_SOURCES_POOL, m_freeSources.data());
+    alDistanceModel(AL_LINEAR_DISTANCE_CLAMPED);
+}
+
+OpenALAudioEngine::~OpenALAudioEngine()
+{
+    m_bufferedSounds.clear();
```

Comment:
IIRC buffers of openAL should be removed earlier than "closing device". I guess comment explaining why it is here will be helpful.